Process Repeaters, Part 1

kaapstorm opened the pull request:
## Technical Summary

At the moment, repeat records are processed independently, and this prevents us from making better decisions about the APIs that we are sending data to. e.g. If a remote server is offline, we will keep trying to send new payloads to it even if all the previous payloads failed.

The approach taken in this PR intends to process repeaters smarter, and also to share repeat record queue workers more fairly across domains. It honors rate limiting by skipping repeaters or domains when they are rate-limited.

Context:
* Jira: [SC-3871](https://dimagi.atlassian.net/browse/SC-3871)
* [Spec](https://docs.google.com/document/d/1axGyCLBJVohz2Fw_JwzpFKwUijDa38TDPEFluzAo9nI/edit)
* Previous (draft) PR https://github.com/dimagi/commcare-hq/pull/34946

This branch implements the feedback given in the draft PR https://github.com/dimagi/commcare-hq/pull/34946 with commits squashed and rebased. :fish: :tropical_fish: :dolphin: 

Almost all new functionality is behind a feature flag. The one change that is not is that [up to once a minute](https://github.com/dimagi/commcare-hq/blob/7e2630ca566cb2c0bc344123a73949b995e4335c/corehq/motech/repeaters/const.py#L16) (less often if there are repeaters being processed) we loop through the results of `RepeaterManager.all_ready()`. ([That happens here](https://github.com/dimagi/commcare-hq/blob/a016d534eea9fa8b2ca1d7c7a9b1e9f0e222595a/corehq/motech/repeaters/tasks.py#L301).)

Enabling the "process_repeaters" feature flag will switch a domain from the current approach to the new approach. The domain can be switched back to the current approach without consequences.

There are two migrations in this PR. I have kept them separate to make it easier to modify the second one. I am looking for confirmation that the indexes in [Index fields used by `RepeaterManager.all_ready()`](https://github.com/dimagi/commcare-hq/pull/35033/commits/e83e925c20d63ee7a6a5b1321094cc6b1d6616b0) are correct.

Currently there are no new Datadog metrics in this PR. I expect that as we test this more / switch domains over, we will add metrics in this PR and follow-up PRs.


## Feature Flag

process_repeaters


## Safety Assurance

### Safety story

* Tested locally and on Staging.
* Behavior is behind a feature flag.


### Automated test coverage

Includes coverage. Please highlight any gaps.


### QA Plan

QA ticket: [QA-7038](https://dimagi.atlassian.net/browse/QA-7038)

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SC-3871]: https://dimagi.atlassian.net/browse/SC-3871?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[QA-7038]: https://dimagi.atlassian.net/browse/QA-7038?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kaapstorm committed:
Add tests

kaapstorm committed:
Add `PROCESS_REPEATERS` toggle

kaapstorm committed:
`process_repeaters()` task

kaapstorm committed:
`get_repeater_lock()`

kaapstorm committed:
`iter_ready_repeater_ids_once()`

kaapstorm committed:
Skip rate-limited repeaters

kaapstorm committed:
`process_repeater()` task

kaapstorm committed:
`Repeater.max_workers` field

snopoke commented on corehq/motech/repeaters/migrations/0015_alter_repeater_is_paused_and_more.py:
Indexes on boolean fields typically don't get used but a useful approach can be to use a partial index to match the query you are trying to optimize. In this case I think it it may work to create a partial index on `next_attempt_at` with a condition for `is_paused=False`.

kaapstorm replied:
Like so? e83e925c20d

snopoke replied:
Yes. Still worth testing that to make sure the query uses it.

snopoke commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -215,6 +226,198 @@ def _process_repeat_record(repeat_record):
+    while True:
+        yielded = False
+        for domain, repeater_id in iter_ready_repeater_ids_once():
+            if not toggles.PROCESS_REPEATERS.enabled(domain, toggles.NAMESPACE_DOMAIN):
```
In a loop like this it's probably better to hold the full list of enabled domains in memory and check access that way. You can refresh the list every few seconds.

kaapstorm replied:
:thinking: I've just checked, and it isn't cached, but somehow I thought it was. A list is a great idea.

snopoke replied:
It's cached way down here: https://github.com/dimagi/commcare-hq/blob/master/corehq/toggles/models.py#L35

You can use `toggles.PROCESS_REPEATERS.get_enabled_domains()` to get the list of domains.

kaapstorm replied:
db093d0eac0 This will refresh the list of domains every time it fetches repeater IDs by domain. It should be equivalent to filtering using the query, but avoid [very big queries](https://github.com/dimagi/commcare-hq/pull/34946#discussion_r1714375956) when a lot of domains are enabled.

kaapstorm replied:
The incremental rollout might be very useful on production. https://github.com/dimagi/commcare-hq/pull/35033/commits/418ed3afe7f853edeeb1160bd025608ce85a0c99

kaapstorm committed:
Index fields used by `RepeaterManager.all_ready()`

kaapstorm committed:
Use quickcache. Prefilter enabled domains.

kaapstorm committed:
Check randomly-enabled domains

kaapstorm committed:
Forward new records for synchronous case repeaters

gherceg commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -215,6 +226,210 @@ def _process_repeat_record(repeat_record):
+
+    def get_task_signature(repeat_record):
+        task_ = {
+            State.Pending: process_pending_repeat_record,
+            State.Fail: process_failed_repeat_record,
```
Just curious why move away from the `process_repeat_record` vs `retry_process_repeat_record`? I like that it is easy to understand the difference whereas these methods have the state context baked into the name, so someone has to understand how state works on repeat records.

gherceg replied:
Ah these are happening in parallel of course, so we can't smoothly transition to continuing to use the existing method names.

gherceg commented on settings.py:
```diff
@@ -628,6 +628,15 @@
+# repeater can use to send repeat records at the same time. This is a
+# guardrail to prevent one repeater from hogging repeat_record_queue
+# workers and to ensure that repeaters are iterated fairly.
+MAX_REPEATER_WORKERS = 144
```
This is a function of # of celery workers right? It is out of scope of this PR, but it would be pretty cool if we could figure out how many workers celery has dedicated to a specific queue and set this to some percentage of that (looks like that is 50% at the moment).

kaapstorm replied:
Yeah, I agree. It would be excellent to be able to calculate this value.

(It is 144 here because that was the number of workers at the time of writing. It was recently doubled.)

gherceg reviewed (commented):
First pass, didn't dig into tests much.

millerdev commented on corehq/motech/repeaters/models.py:
```diff
@@ -268,6 +268,7 @@ class Repeater(RepeaterSuperProxy):
     is_paused = models.BooleanField(default=False)
     next_attempt_at = models.DateTimeField(null=True, blank=True)
     last_attempt_at = models.DateTimeField(null=True, blank=True)
+    max_workers = models.IntegerField(default=0)
```
What do you think of spinning this feature off in a separate PR (if we decided it's needed)? We have a lot of knobs to turn already, and I wonder if we'll need this one?

kaapstorm replied:
We needed this one years ago. OpenMRS integrations involve Events and Observations, and it is _really useful_ if we can send repeat records in the order in which their forms were submitted. More generally, setting `max_workers` to 1 allows us to ensure that repeat records are sent chronologically.

And setting it to [all of the workers] is an insurance policy for this PR to ensure that we can still handle the highest volume for repeaters that need it. I'd sleep easier if this is included.

millerdev commented on corehq/motech/repeaters/migrations/0015_alter_repeatrecord_state_and_more.py:
Similar question about putting migrations in a separate PR.

millerdev commented on corehq/motech/repeaters/models.py:
```diff
@@ -1327,3 +1381,11 @@ def domain_can_forward(domain):
     )
+
+
+@quickcache(['domain'], timeout=60)
```
Can we remove this and only add it later if we find out that it's needed?

kaapstorm replied:
> or is the idea that with a short timeout, it isn't a big deal if we operate with stale data for less than a minute?

Yes, that.

> Can we remove this and only add it later if we find out that it's needed?

Could you say more about this suggestion? We call `domain_can_forward_now()` many times a second, and it calls `toggles.PAUSE_DATA_FORWARDING.enabled()`. The consequence of caching it is that it takes under a minute to pause or unpause data forwarding for a domain, but it saves what looks like an expensive function call. I'm not sure what you mean by "if ... it's needed". In terms of performance, it seems better to cache it than not to cache it. What am I not considering? What is the downside of caching it?

millerdev replied:
> I'm not sure what you mean by "if ... it's needed".

I meant after we have determined that performance is bad without caching, which would indicate that we need it.

> We call `domain_can_forward_now()` many times a second, and it calls `toggles.PAUSE_DATA_FORWARDING.enabled()`. The consequence of caching it is that it takes under a minute to pause or unpause data forwarding for a domain, but it saves what looks like an expensive function call.

This made me think you have measured performance and determined that we need caching. If that's true, then you can ignore my request.

> In terms of performance, it seems better to cache it than not to cache it.

However, this part made me wonder if we just added caching because "it seems better" (without measuring performance first).

> What is the downside of caching it?

Complexity. Cached values can be hard to reason about, especially in asynchronous/concurrent systems.

kaapstorm replied:
4955ef4a9628c5238956ea07dc6e592cc73b65a6

millerdev reviewed (commented):
Pair reviewed with @gherceg.

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -1171,6 +1171,9 @@ def attempt_forward_now(self, *, is_retry=False, fire_synchronously=False):
         )
 
+        if toggles.PROCESS_REPEATERS.enabled(self.domain, toggles.NAMESPACE_DOMAIN):
+            return
```
If by "gracefully" you mean "will all repeat records be processed when the toggle is disabled?", then yes.

But if by "gracefully" you mean "will Datadog show the repeat records as not overdue?", then no. Datadog will show the repeat records as overdue.

Alternatively, when we apply a back-off to a repeater, we could also update all its pending and failed repeat records too. I considered this, but it felt like a lot of churn on the repeat records table. I thought a better approach would be to use a new metric for Datadog to gauge when the repeat records queue is getting backed up.

kaapstorm replied:
30d4a6fac0d

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -1327,3 +1330,10 @@ def domain_can_forward(domain):
     )
+
+
+def domain_can_forward_now(domain):
```
I'll add docstrings. We can't combine them because they are used differently. -- When `domain_can_forward()` is False, we don't register a repeat record.

kaapstorm replied:
d1119bbec47

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -256,17 +260,29 @@ def iter_ready_repeater_ids_forever():
-            yield repeater.domain, repeater.repeater_id
+
+            lock = get_repeater_lock(repeater.repeater_id)
+            lock_token = uuid.uuid1().hex  # The same way Lock does it
```
> Why use `uuid1` here rather than the more standard `uuid4`?

This is what the comment refers to. Sorry it wasn't more clear. I'll make it clearer.

I set the value of `lock_token` [the same way that `redis.lock.Lock` sets its lock token](https://github.com/redis/redis-py/blob/master/redis/lock.py#L198).

> Possibly add a comment to explain why it is needed?

Will do. (Redis uses the lock token to determine the ownership of the lock. If a process knows the value of the token, then that process is considered the owner of the lock, and it is allowed to release the lock without first acquiring it.)

kaapstorm replied:
d1119bbec47

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -256,17 +260,29 @@ def iter_ready_repeater_ids_forever():
             return
 
 
+def get_repeater_lock(repeater_id):
```
Yes. When you try to acquire the lock returned by `get_redis_lock()`, you get the following TypeError:

```
File /srv/share/src/dimagi/commcare-hq/.venv/lib/python3.9/site-packages/redis/lock.py:207, in Lock.acquire(self, sleep, blocking, blocking_timeout, token)
    205     stop_trying_at = mod_time.monotonic() + blocking_timeout
    206 while True:
--> 207     if self.do_acquire(token):
    208         self.local.token = token
    209         return True

File /srv/share/src/dimagi/commcare-hq/.venv/lib/python3.9/site-packages/redis/lock.py:223, in Lock.do_acquire(self, token)
    221 else:
    222     timeout = None
--> 223 if self.redis.set(self.name, token, nx=True, px=timeout):
    224     return True
    225 return False

File /srv/share/src/dimagi/commcare-hq/.venv/lib/python3.9/site-packages/django_redis/cache.py:29, in omit_exception.<locals>._decorator(self, *args, **kwargs)
     26 @functools.wraps(method)
     27 def _decorator(self, *args, **kwargs):
     28     try:
---> 29         return method(self, *args, **kwargs)
     30     except ConnectionInterrupted as e:
     31         if self._ignore_exceptions:

File /srv/share/src/dimagi/commcare-hq/.venv/lib/python3.9/site-packages/django_redis/cache.py:81, in RedisCache.set(self, *args, **kwargs)
     79 @omit_exception
     80 def set(self, *args, **kwargs):
---> 81     return self.client.set(*args, **kwargs)

TypeError: set() got an unexpected keyword argument 'px'
```

`django_redis` appears to use an incompatible client. Using `redis.lock.Lock` with `get_redis_connection()` resolved that.

kaapstorm replied:
:thinking: After repeating the exact steps I wrote down to reproduce this error, I no longer get it. I have no explanation.

Switched to using `get_redis_lock()` and added a test for the way we are using the lock token. 03b26cfb41d

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -372,7 +385,7 @@ def set_next_attempt(self):
-    def reset_next_attempt(self):
+    def reset_backoff(self):
         if self.last_attempt_at or self.next_attempt_at:
             self.last_attempt_at = None
```
Backoff uses `last_attempt_at` to determine the duration of the next backoff interval.

Repeat records implement exponential backoff by multiplying the last interval by 3, and this code has preserved that behavior.

We set `last_attempt_at` to `None` so that the next time we need to back off, we know it is the initial interval and should be set it to the minimum. This matches [the way we use `RepeatRecord.last_checked` when setting `RepeatRecord.next_check`](https://github.com/dimagi/commcare-hq/blob/85bbfa337942042f5539d143b6b6238d44eb9a1f/corehq/motech/repeaters/models.py#L1106) for exponential backoff currently.

millerdev replied:
Is that explained somewhere in a comment or docstring? If not maybe one could be added?

kaapstorm replied:
b70fc523d63

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -215,6 +226,210 @@ def _process_repeat_record(repeat_record):
+    if not process_repeater_lock.acquire(blocking=False):
+        return
+    try:
+        for domain, repeater_id, lock_token in iter_ready_repeater_ids_forever():
+            process_repeater.delay(domain, repeater_id, lock_token)
```
> Do we need to be checking how long this task has been running for and exit if we are about to exceed the `process_repeater_lock` timeout value?

`process_repeater()` should take a maximum of five minutes, if that repeater's requests are timing out. The lock timeout is three hours.

I expect we would only exceed the timeout if Celery was killed and the `update_repeater()` task was never called. So the timeout is intended to prevent a repeater from being locked forever.

I think it could be useful to use Datadog to monitor how long the task takes, but probably only to identify problematic endpoints, not to exit the task.

gherceg replied:
I may not be understanding, but I'm referring to the `process_repeaters` task and the `process_repeater_lock` with the 24 hour timeout.

kaapstorm replied:
Sorry for my confusion. I didn't read your question properly.

I wondered about that timeout too, because there is no problem if this loop just keeps going for more than 24 hours _as long as_ new repeat records are being sent in a reasonable time.

What do you think of not having a timeout?

kaapstorm replied:
4c4189661f8

kaapstorm committed:
Add explanatory docstrings and comments

kaapstorm committed:
get_redis_lock() ... acquire(): No TypeError ?!

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -255,24 +255,61 @@ def iter_ready_repeater_ids_forever():
+
+    def iter_domain_repeaters(dom):
+        try:
+            rep_id = repeater_ids_by_domain[dom].pop(0)
```
No, it's not important, it just made the tests a little more readable. I didn't realize that `pop(0)` had performance implications -- I thought Python would just move the pointer to the start of the list one item up, but you are right (and this was quite fun to verify). I'll change it.

kaapstorm replied:
de27ba0da31

kaapstorm committed:
Drop unnecessary `iter_domain_repeaters()`

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -255,24 +255,61 @@ def iter_ready_repeater_ids_forever():
+        for domain in list(repeater_ids_by_domain.keys()):
+            try:
+                repeater_id = next(iter_domain_repeaters(domain))
+            except StopIteration:
```
de27ba0da31

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -255,24 +255,61 @@ def iter_ready_repeater_ids_forever():
+                del repeater_ids_by_domain[domain]
+                continue
+            else:
+                yield domain, repeater_id
```
de27ba0da31

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -322,7 +323,99 @@ def get_repeater_lock(repeater_id):
+                return None
+            with timer('fire_timing') as fire_timer:
+                state_or_none = repeat_record.fire(timing_context=fire_timer)
+            report_repeater_usage(
```
:thinking: Probably not. @gherceg ?

gherceg replied:
So given request related exceptions are handled in `repeater.fire_for_record` (which gets called by `repeat_record.fire`), it should be true that if none is returned here, it is because an exception was hit when getting the payload or because the repeat record's state is None which I don't think is possible. However it doesn't hurt to report the usage, and we currently _do_ report the timing of both fetching the payload and sending the request.

All that to say, it seems fine to leave it as is for now to remain consistent with how we report timing for the rate limiter. If we decide to reduce the scope of what usage is reported to the request specifically, we can update this code then.

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -322,7 +323,99 @@ def get_repeater_lock(repeater_id):
+        if any(s == State.Success for s in repeat_record_states):
+            # At least one repeat record was sent successfully. The
+            # remote endpoint is healthy.
+            repeater.reset_backoff()
```
I don't think so. If an API always responds slowly then we'll end up only sending `repeater.num_workers` repeat records [once every seven days](https://github.com/dimagi/commcare-hq/blob/0db6a6c4f4ca90d3650f7d7beaf753042b97df8f/corehq/motech/repeaters/const.py#L7) (the maximum back-off interval). That seems wrong.

Compare that to the current behavior, where we will always try to send when we register a repeat record, even if every single repeat record has failed before.

Backing off from a healthy endpoint is a scenario that really worries me about this new process, because it could mean that no payloads go through for long intervals, and that can have strong negative consequences.

kaapstorm committed:
Don't quickcache `domain_can_forward_now()`

kaapstorm committed:
Migration to create indexes concurrently

orangejenny committed:
Merge branch 'master' into nh/iter_repeaters_1

kaapstorm committed:
Add comment

kaapstorm committed:
Don't squash BaseExceptions

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -1168,7 +1181,9 @@ def fire(self, force_send=False, timing_context=None):
                 self.handle_payload_error(str(e), traceback_str=traceback.format_exc())
-                raise
+            finally:
+                return self.state
```
7e65b3b0661

kaapstorm committed:
Drop timeout for `process_repeater_lock`.

kaapstorm committed:
Add metric for monitoring health

kaapstorm committed:
Merge branch 'master' into nh/iter_repeaters_1

kaapstorm committed:
Resolve migration conflict, fix index

kaapstorm committed:
Fix metric

kaapstorm commented on corehq/motech/repeaters/migrations/0016_add_indexes.py:
```diff
@@ -0,0 +1,62 @@
+                ),
+                migrations.RunSQL(
+                    sql="""
+                    CREATE INDEX CONCURRENTLY "next_attempt_at_partial_idx"
+                    ON "repeaters_repeater" ("next_attempt_at")
+                    WHERE (NOT "is_deleted" AND NOT "is_paused");
```
I analyzed the query used by `Repeater.objects.get_all_ready_ids_by_domain()`, and this is the result on Staging:
```
> EXPLAIN ANALYZE SELECT "repeaters_repeater"."domain", "repeaters_repeater"."id_"
  FROM "repeaters_repeater"
    INNER JOIN "repeaters_repeatrecord"
    ON ("repeaters_repeater"."id_" = "repeaters_repeatrecord"."repeater_id_")
  WHERE (
    NOT "repeaters_repeater"."is_deleted"
    AND NOT "repeaters_repeater"."is_paused"
    AND (
      "repeaters_repeater"."next_attempt_at" IS NULL
      OR "repeaters_repeater"."next_attempt_at" <= '2024-10-19 22:11:36.310082'
    )
    AND "repeaters_repeatrecord"."state" IN (1, 2)
  );

                                                                                      QUERY PLAN                                                                                      
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=8423.11..141317.33 rows=52115 width=30) (actual time=61.472..1547.920 rows=638862 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Hash Join  (cost=7423.11..135105.83 rows=21715 width=30) (actual time=56.214..1224.831 rows=212954 loops=3)
         Hash Cond: (repeaters_repeatrecord.repeater_id_ = repeaters_repeater.id_)
         ->  Parallel Bitmap Heap Scan on repeaters_repeatrecord  (cost=7289.92..134244.51 rows=276830 width=16) (actual time=52.249..1077.808 rows=223024 loops=3)
               Recheck Cond: (state = ANY ('{1,2}'::integer[]))
               Heap Blocks: exact=16852
               ->  Bitmap Index Scan on repeaters_repeatrecord_state_8055083b  (cost=0.00..7123.82 rows=664393 width=0) (actual time=44.255..44.256 rows=669452 loops=1)
                     Index Cond: (state = ANY ('{1,2}'::integer[]))
         ->  Hash  (cost=131.11..131.11 rows=167 width=30) (actual time=0.713..0.722 rows=163 loops=3)
               Buckets: 1024  Batches: 1  Memory Usage: 18kB
               ->  Bitmap Heap Scan on repeaters_repeater  (cost=5.84..131.11 rows=167 width=30) (actual time=0.089..0.660 rows=163 loops=3)
                     Filter: ((NOT is_deleted) AND (NOT is_paused) AND ((next_attempt_at IS NULL) OR (next_attempt_at <= '2024-10-19 22:11:36.310082+00'::timestamp with time zone)))
                     Rows Removed by Filter: 40
                     Heap Blocks: exact=67
                     ->  Bitmap Index Scan on repeaters_repeater_is_deleted_08441bf0  (cost=0.00..5.80 rows=203 width=0) (actual time=0.053..0.059 rows=253 loops=3)
                           Index Cond: (is_deleted = false)
 Planning Time: 0.874 ms
 Execution Time: 1623.954 ms
(20 rows)
```

The "next_attempt_at_partial_idx" was chosen to match the filter in that query plan, but Postgres is not using the index.

Should we drop it, change it, or leave it?

millerdev replied:
How much data is present on staging? If not much then those two indexes may have had very similar outcomes and the choice of which to use may not have been very meaningful.

I'd check the same or a similar query on prod to see if you get the same query plan.

Do we need both indexes? Would it break other queries if we added the partial condition `NOT "is_deleted" AND NOT "is_paused"` on `repeaters_repeatrecord_state_8055083b` and eliminated the `next_attempt_at_partial_idx` index?

kaapstorm replied:
Interesting! Thank you!

kaapstorm replied:
Very nice. It uses the index.
```
QUERY PLAN
----------
 Gather  (cost=1078.24..298737.78 rows=91675 width=30) (actual time=1860.938..7291.760 rows=952533 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Parallel Append  (cost=78.24..288570.28 rows=91675 width=30) (actual time=1103.767..5783.887 rows=317511 loops=3)
         ->  Hash Join  (cost=132.62..143624.77 rows=31897 width=30) (actual time=1.539..2456.512 rows=317502 loops=3)
               Hash Cond: (repeaters_repeatrecord.repeater_id_ = repeaters_repeater.id_)
               ->  Parallel Seq Scan on repeaters_repeatrecord  (cost=0.00..142389.57 rows=419192 width=16) (actual time=1.080..2293.795 rows=327791 loops=3)
                     Filter: (state = ANY ('{1,2}'::integer[]))
                     Rows Removed by Filter: 1776517
               ->  Hash  (cost=130.60..130.60 rows=162 width=30) (actual time=0.357..0.359 rows=153 loops=3)
                     Buckets: 1024  Batches: 1  Memory Usage: 17kB
                     ->  Bitmap Heap Scan on repeaters_repeater  (cost=5.84..130.60 rows=162 width=30) (actual time=0.067..0.295 rows=153 loops=3)
                           Filter: ((NOT is_deleted) AND (NOT is_paused) AND (next_attempt_at IS NULL))
                           Rows Removed by Filter: 50
                           Heap Blocks: exact=67
                           ->  Bitmap Index Scan on repeaters_repeater_is_deleted_08441bf0  (cost=0.00..5.80 rows=203 width=0) (actual time=0.044..0.045 rows=253 loops=3)
                                 Index Cond: (is_deleted = false)
         ->  Hash Join  (cost=78.24..143570.39 rows=6301 width=30) (actual time=1655.248..4904.097 rows=14 loops=2)
               Hash Cond: (repeaters_repeatrecord_1.repeater_id_ = repeaters_repeater_1.id_)
               ->  Parallel Seq Scan on repeaters_repeatrecord repeaters_repeatrecord_1  (cost=0.00..142389.57 rows=419192 width=16) (actual time=0.043..4756.595 rows=491687 loops=2)
                     Filter: (state = ANY ('{1,2}'::integer[]))
                     Rows Removed by Filter: 2664776
               ->  Hash  (cost=77.84..77.84 rows=32 width=30) (actual time=39.663..39.664 rows=10 loops=2)
                     Buckets: 1024  Batches: 1  Memory Usage: 9kB
                     ->  Bitmap Heap Scan on repeaters_repeater repeaters_repeater_1  (cost=4.39..77.84 rows=32 width=30) (actual time=39.626..39.647 rows=10 loops=2)
                           Recheck Cond: ((next_attempt_at <= '2024-10-19 22:11:36.310082+00'::timestamp with time zone) AND (NOT is_deleted) AND (NOT is_paused))
                           Heap Blocks: exact=7
                           ->  Bitmap Index Scan on next_attempt_at_partial_idx  (cost=0.00..4.38 rows=32 width=0) (actual time=39.577..39.577 rows=10 loops=2)
                                 Index Cond: (next_attempt_at <= '2024-10-19 22:11:36.310082+00'::timestamp with time zone)
 Planning Time: 3.019 ms
 Execution Time: 7411.459 ms
(31 rows)
```

Out of curiosity, I'll try a composite index and see what difference it makes. Either way, this feels like a big step in the right direction.

kaapstorm replied:
With a composite index, planning time is 2.5 ms faster, but execution time is 363 ms slower. The partial index wins.

kaapstorm replied:
I found some changes that made a significant difference. efc4dde3b96803301a2a5ec888b15c38c8941623

The details are in the commit message:

> After some analyzing, it turns out that the query is between two and three times as fast using a partial index on `RepeatRecord.repeater_id` where `state` is Pending or Failed. Performance is also improved using a partial index with `Repeater.is_deleted` instead of a normal B-tree index. When these two indexes are used, the `next_attempt_at_partial_idx` is not used.

I've dropped the `next_attempt_at_partial_idx` because of that, and changed the indexes for `Repeater.state` and `Repeater.is_deleted` to be partial.

millerdev replied:
I think `Parallel Seq Scan on repeaters_repeatrecord` is going to be a problem if that query plan is chosen on prod. In comparison, `Index Scan on repeaters_repeater_...` is inconsequential because there are only ~2000 (tiny number) repeaters whereas there are millions of repeat records.

kaapstorm replied:
Just to close the loop here: I tested the RepeatRecord.state index on Prod, and Postgres uses it for both subqueries. I tested on Friday afternoon, and the query took around 27 seconds.

kaapstorm replied:
I think the benefits of this PR are big enough for us to iterate on step-wise improvements after it is merged. One improvement I think we should consider is replacing the `repeaters_repeatrecord` table with nested partitioning / subpartitioning. Something like:

```
CREATE TABLE repeaters_repeatrecord_p (
  -- ...
) PARTITION BY LIST (state);

CREATE TABLE repeaters_repeatrecord_ready
  PARTITION OF repeaters_repeatrecord_p
  FOR VALUES IN (1, 2);

CREATE TABLE repeaters_repeatrecord_done
  PARTITION OF repeaters_repeatrecord_p
  FOR VALUES IN NOT (1, 2)
  PARTITION BY RANGE (registered_on);

CREATE TABLE repeaters_repeatrecord_2010
  PARTITION OF repeaters_repeatrecord_done
  FOR VALUES FROM ('2010-01-01') TO ('2011-01-01');

-- etc.
```

This will improve queries on repeat records ready to be sent, and also allow us to drop completed repeat records after some expiry (TBD).

kaapstorm committed:
Change indexes

After some analyzing, it turns out that the query is between two and
three times as fast using a partial index on RepeatRecord.repeater_id
where state is pending or failed. Performance is also improved using a
partial index with Repeater.is_deleted instead of a normal B-tree
index. When these two indexes are used, the next_attempt_at_partial_idx
is not used.

kaapstorm committed:
Add one more index. Use UNION ALL queries.

kaapstorm committed:
Don't report attempt too soon

kaapstorm committed:
Add metrics

kaapstorm committed:
Improve backoff logic

kaapstorm committed:
Update comments

kaapstorm committed:
Show "Next attempt at" in Forwarders page

kaapstorm committed:
Merge branch 'master' into nh/iter_repeaters_1

kaapstorm committed:
Fixes migration

kaapstorm committed:
Merge remote-tracking branch 'origin/master' into nh/iter_repeaters_1

kaapstorm committed:
Add comment on other `True` return value

kaapstorm committed:
Count repeater backoffs

kaapstorm committed:
Add documentation

millerdev commented on corehq/motech/repeaters/migrations/0016_add_indexes.py:
```diff
@@ -12,49 +12,61 @@ class Migration(migrations.Migration):
+                migrations.AddIndex(
+                    model_name="repeatrecord",
+                    index=models.Index(
+                        condition=models.Q(("state__in", (1, 2))),
```
I wonder what are the computational differences in Postgres for `state IN (1, 2)` vs `state = 1 OR state = 2`? I hope they are virtually the same, but it would be a shame if `IN (...)` does something complicated with a loop.

millerdev commented on corehq/motech/repeaters/migrations/0016_add_indexes.py:
```diff
@@ -12,49 +12,61 @@ class Migration(migrations.Migration):
             ],
 
             database_operations=[
+                # Drop `Repeater.id_deleted` index
```
```suggestion
                # Drop `Repeater.is_deleted` index
```

millerdev commented on corehq/motech/repeaters/tests/test_tasks.py:
```diff
@@ -558,3 +559,65 @@ def test_update_repeater_does_nothing_on_none(self, mock_get_repeater, __):
+            registered_at=five_minutes_ago,
+        )
+        wait_duration = _get_wait_duration_seconds(repeat_record)
+        self.assertEqual(wait_duration, 300)
```
Could there be a race condition (resulting in a flaky test) here if something took an extra second or two (like garbage collection) between when `repeat_record` is created and when `_get_wait_duration_seconds` is run? Seems unlikely, but not impossible. Maybe consider using something like `freezegun.freeze_time`?

The same applies to other tests in this class as well.

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -1,3 +1,76 @@
+round-robin fashion, cycling through the domains. It does this so that:
+
+* Domains and repeaters are not rate-limited unnecessarily.
+* Remote APIs are not unintentionally DDoS-attacked by CommCare HQ.
```
nit: Technically this would be a "DoS", not a "DDoS".

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -1,3 +1,76 @@
+set per repeater) and spawns tasks to send them in parallel. The results
+of all the send attempts are passed to ``update_repeater()``. If all the
+send attempts failed, the **repeater** (not the repeat record) is backed
+off. If any send attempts succeeded, the backoff is reset.
```
Imagine a scenario where the first attempt usually succeeds, but subsequent attempts usually fail. I think we would ideally want to reduce the batch size for such a fragile repeater endpoint, but I assume we have no automated way of doing that right now. Do we have any way of detecting that scenario?

kaapstorm replied:
> Do we have any way of detecting that scenario?

Not yet, but once this PR is merged, we will be tying up a few loose ends, and this is one of them. If we get a 429 response, we will automatically reduce `Repeater.num_workers`. This doesn't specifically take your example into account, but if the remote API uses 429 responses, then it will handle, say, the first 5 attempts succeeding and the last two failing.

kaapstorm committed:
Use less aggressive backoff for repeaters

kaapstorm committed:
Add TODOs

kaapstorm committed:
Merge branch 'master' into nh/iter_repeaters_1

kaapstorm committed:
Add tests

kaapstorm committed:
Return one row per repeater

kaapstorm committed:
Drop RepeaterManager.all_ready()

kaapstorm committed:
Refactor queries to count RepeatRecords

kaapstorm committed:
Drop Datadog tag

kaapstorm committed:
Tweak metrics

kaapstorm committed:
Merge branch 'master' into nh/iter_repeaters_1

millerdev commented on corehq/motech/repeaters/models.py:
```diff
@@ -233,22 +235,44 @@ def __new__(cls, *args, **kwargs):
+        return results
+
+    def get_all_ready_next_attempt_null(self):
+        # See `_all_ready_next_attempt_now()`. Splitting the queries
```
```suggestion
        # See `get_all_ready_next_attempt_now()`. Splitting the queries
```

kaapstorm committed:
Set much shorter timeout on repeater lock

kaapstorm committed:
Fix comments

kaapstorm committed:
Nit

kaapstorm committed:
Fix test

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -215,9 +297,265 @@ def _process_repeat_record(repeat_record):
+            # lock and a different process can release it. This lock
+            # will be released by the `update_repeater()` task.
+            lock_token = uuid.uuid1().hex
+            if lock.acquire(blocking=False, token=lock_token):
+                yielded = True
+                yield domain, repeater_id, lock_token
+            else:
+                metrics_counter(
+                    'commcare.repeaters.process_repeaters.repeater_locked',
+                    tags={'domain': domain},
+                )
```
I've figured out what I'm missing:

The [`process_repeaters()` task](https://github.com/dimagi/commcare-hq/blob/b1dcbbd06d7b6bb6ed5c0086e90dfe39761937fe/corehq/motech/repeaters/tasks.py#L304) (note the final "s") spawns an asynchronous [`process_repeater()` task](https://github.com/dimagi/commcare-hq/blob/b1dcbbd06d7b6bb6ed5c0086e90dfe39761937fe/corehq/motech/repeaters/tasks.py#L418) for every repeater. This takes next to no time at all. Then it loops again, and of course all the repeaters are still locked, because not a single sent payload has got a reply yet. So it exits and waits a minute.

kaapstorm committed:
Don't spawn a subtask for `process_repeater()`

kaapstorm committed:
Management command to expire lock

kaapstorm committed:
An alternate approach to iterating until complete

kaapstorm committed:
Clean up tests

kaapstorm committed:
Update docstrings

kaapstorm commented:
### Findings

While working on this change, I found a more performant way to loop through repeaters. I've outlined the approach in the module docstring of `corehq/motech/repeaters/tasks.py`. Or you can look at the code of `process_repeaters()`.

The following Datadog screenshot illustrates how this process works.

![Screenshot from 2025-01-11 23-27-54](https://github.com/user-attachments/assets/ad703117-b2be-463c-98a0-95216af0e932)

The "Repeat records ready" graph shows that about 1,000 repeat records are being processed, then a few hundred new repeat records are added, and then they are all processed.

The "Repeat records by domain" graph shows that at first all repeat records are from the "nhooper" domain. The new repeat records have been added to four other domains. They are processed in a round-robin way where repeat records are pulled from all five domains fairly. Once the repeat records from the four domains have all been sent, the remainder of the repeat records are sent from the "nhooper" domain.

The "Iterated all repeaters once" graph shows this from the perspective of looping over the repeaters with repeat records ready to be sent. Initially, only one repeater has repeat records ready to be sent, and so the loop is run through over and over rapidly. After new repeat records have been added to many repeaters across five domains, the loop takes much longer to loop through all of them. And after the new repeat records have been sent, once again `process_repeaters()` is looping over only one repeater.

The "Repeat record wait times" graph shows the age of the repeat records that are being sent. The first repeater, in the "nhooper" domain, has the oldest repeat records ("lt_466560s" is less than 5 days).

Lastly, the gaps in the "Repeat records by domain" and "Repeat record wait times" graphs are interesting. They appear to be when the repeat record queue workers are being used by the old process.


### Are we there yet?

QA is complete.

Metrics have been added for visibility into the process.

Further testing built off the work that QA had done, in order to improve how this new process deals with backing off.

And then more testing resulted in an improvement to the way in which we keep processing repeaters until all repeat records have been processed.

I think we're there.


### Follow-ups

* Add the ability to respect rate-limiting responses from remote APIs.

* Add the ability to merge payloads (e.g. merge case updates, or UCR data source changes).

kaapstorm commented:
### A note to reviewers

If you have reviewed this PR before, and you have a rough idea of what this new process does, then the important changes since you read this last are probably just the last three commits:
* [An alternate approach to iterating until complete](https://github.com/dimagi/commcare-hq/pull/35033/commits/964ae323e58d1ff8fedcbf802cfe0ecb11e57d9f)
* [Clean up tests](https://github.com/dimagi/commcare-hq/pull/35033/commits/b801159fdfedd765fadf530f3ae79c308da02c3b)
* [Update docstrings](https://github.com/dimagi/commcare-hq/pull/35033/commits/90bf28144bff7308e8d68bbad9562c9fe6a0d987)

@gherceg @millerdev

kaapstorm committed:
Improve wording

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -322,50 +321,35 @@ def process_repeaters():
+                tasks.append(process_repeater_chord(domain, repeater_id))
+
+            if tasks:
+                result = group(*tasks).apply_async()
```
The maximum possible size is the number of repeaters in the environment if all of the repeaters are in use. In practice, though, the number of repeaters in use at any moment will be a fraction of that.

On Staging I tested with 60 repeaters concurrently in use, across 4 repeat record queue workers.

I think Prod currently has 144 or 288 repeat record queue workers. I'm not sure of the average number of repeaters in use at any moment.

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -322,50 +321,35 @@ def process_repeaters():
+                # https://docs.celeryq.dev/en/stable/userguide/tasks.html#avoid-launching-synchronous-subtasks
+                # but in this situation it is safe because the
+                # `process_repeater()` tasks use a different queue.
+                result.get(disable_sync_subtasks=False)
```
> Would it be possible to restructure this to avoid waiting on this result?

I could not think of a way, but I would be delighted if anyone has suggestions. (Copilot was no help either, but maybe I was asking wrong.)

What I was hoping for was something similar to [`Executor.map()`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.map) that could _lazily_ map repeater IDs or `process_repeater()` tasks onto Celery workers as the workers free up. The missing piece in the Celery puzzle is that I couldn't find a way **to wait** for a worker to free up in order to pass it the next repeater ID or task.

So what you end up with is the problem that I encountered with my first approach: All of the repeaters are locked as the first iteration of `iter_ready_repeater_ids()` is consumed. I couldn't find a way to wait for the a worker to free up in order to pass it the next ID or task. And so the function that is spawning the tasks will exit.

It is faster to wait for all of the workers to complete (my second approach) than to exit and wait for `process_repeaters()` to run again (my first approach).

A third approach could be to reintroduce the repeater lock, but to _block_ on acquiring the lock, instead of to skip locked repeaters. I expect that performance will be the same as waiting for tasks to finish, because you'll still end up waiting for the slowest repeater.

kaapstorm replied:
What do you think of 03197dc9462?

kaapstorm committed:
Wait for random task instead of all tasks

kaapstorm committed:
Lint

kaapstorm committed:
`MeteredLock` supports `reacquire()`

kaapstorm committed:
`update_repeater()` calls `process_repeater()`

kaapstorm commented:
TLDR: On the whole, I think the (new) new approach is better.

Here is the breakdown:

#### Before
![image](https://github.com/user-attachments/assets/c6b0266f-d14e-4f01-90d5-1e1912030b80)

In the example above, if you look from about 15:48 to 15:53, the round-robin approach works well when all 5 domains have repeat records ready to be sent. But there are times when workers are idle, and the downward gradient on "Repeat records ready" is not as steep as below.

#### After
![image](https://github.com/user-attachments/assets/659f6a0e-86e6-4deb-a586-ce8d816f3d8d)

In this example, it takes longer to start processing the repeaters for the "nhooper" and "repeat-records-2" domains, because we check every 5 minutes instead of every minute. When we do start processing them, their tasks dominate for a little while (18:33 to 18:34) and then Celery flips back to the tasks for the first three domains, and it cycles like that until about 18:37 when the first three domains run out of repeat records.

However, there are no idle workers, and the gradient on "Repeat records ready" is steeper. The process is also fair enough for no domain to be neglected, and no domain to hog resources. It seems clear that this is a better approach.

millerdev reviewed (approved):
I noticed a bunch of tests were [removed](https://github.com/dimagi/commcare-hq/pull/35033/commits/964ae323e58d1ff8fedcbf802cfe0ecb11e57d9f#diff-fab51d3577abdff2fa471823529033021051e561e855770933b0b5514b3a9fd9) [recently](https://github.com/dimagi/commcare-hq/pull/35033/commits/b801159fdfedd765fadf530f3ae79c308da02c3b). Were they not useful tests?

kaapstorm committed:
Add comment for lock.reacquire()

kaapstorm committed:
Encapsulate locking.

kaapstorm committed:
Encapsulate locking in a class

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -546,7 +509,11 @@ def update_repeater(repeat_record_states, repeater_id, lock_token):
         lock.local.token = lock_token
-        lock.release()
+        if more:
+            lock.reacquire()
```
9e8c7eef03f

kaapstorm commented on corehq/util/metrics/tests/test_lockmeter.py:
```diff
@@ -142,6 +142,22 @@ def test_acquire_untracked(self):
+        fake = FakeLock(timeout=-1)
+        lock = MeteredLock(fake, "test", track_unreleased=False)
+        with patch("corehq.util.metrics.lockmeter.tracer") as tracer:
+            lock.reacquire()
```
This test is based on about a dozen other tests in this class. If I assert this line, I should probably assert the corresponding line in all the other tests. But this test's assertion would fail if `lock.reacquire()` had failed (like the assertions in the other tests) so I think it's okay for the tests to skip an assertion on this line.

kaapstorm commented:
> I noticed a bunch of tests were [removed](https://github.com/dimagi/commcare-hq/pull/35033/commits/964ae323e58d1ff8fedcbf802cfe0ecb11e57d9f#diff-fab51d3577abdff2fa471823529033021051e561e855770933b0b5514b3a9fd9) [recently](https://github.com/dimagi/commcare-hq/pull/35033/commits/b801159fdfedd765fadf530f3ae79c308da02c3b). Were they not useful tests?

`TestIterReadyRepeaterIDsForever` tested a function that no longer exists. And `TestProcessRepeater` was an ugly (imho) high-level test that is already covered in lower-level tests.

kaapstorm commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -546,7 +509,11 @@ def update_repeater(repeat_record_states, repeater_id, lock_token):
-        lock.release()
+        if more:
+            lock.reacquire()
+            process_repeater(repeater, lock_token)
```
I might not be imagining exactly what you have in mind, but here is what I think you mean, with a small change. When `update_repeater()` calls `process_repeater()` it already has the repeater instance, so I am hoping to avoid fetching it again.

It's appealing to put the lock-related code in one function, but I noticed that the code to be executed if `lock_token` is set only has one line in common with the code if `lock_token` is `None`. I ended up with encapsulating the locking functionality across two functions. I'm not sure that it's much of an improvement, but I think it's a bit of an improvement: https://github.com/dimagi/commcare-hq/pull/35033/commits/ebec20dd4ef55664faafc235598b426699167f50

Then I took a second pass at encapsulating the locking functionality, and I moved it into a class: https://github.com/dimagi/commcare-hq/pull/35033/commits/09d757ef3de3da2250281252b345dfd520443118 (I vacillated on whether to call it "RepeaterLock" or "RepeaterLockUtil", but I don't like "Util" (or "Helper").)

kaapstorm commented on corehq/tests/pytest_plugins/redislocks.py:
```diff
@@ -48,6 +49,8 @@ class TestLock:
     lock = attr.ib(repr=False)
     timeout = attr.ib()
 
+    local = Mock()
```
This change is to allow setting `lock.local.token = lock_token`.

millerdev replied:
Might be useful to add a comment so future readers of this code understand why it's there. Can be done in a separate PR after this one is merged.

kaapstorm replied:
6e803b154b9

kaapstorm committed:
Management command to set `Repeater.max_workers`.

millerdev reviewed (commented):
This is looking good. I made a few suggestions, which I think can be grouped together in a single commit. However, I think some tests will need to be updated if the suggestions are accepted.

kaapstorm committed:
Add a comment and a couple of tests

kaapstorm committed:
Simplify RepeaterLock

kaapstorm commented:
These are great suggestions, thanks.
